File: DIP/DIP_009/Python/DIP_009_line_detector.py
import numpy as np
import cv2

# Draw the lines represented in the hough accumulator on the original image
def drawhoughLinesOnImage(image, houghLines):
    for line in houghLines:
        for rho,theta in line:
            a = np.cos(theta)
            b = np.sin(theta)
            x0 = a*rho
            y0 = b*rho
            x1 = int(x0 + 1000*(-b))
            y1 = int(y0 + 1000*(a))
            x2 = int(x0 - 1000*(-b))
            y2 = int(y0 - 1000*(a))
            
            cv2.line(image,(x1,y1),(x2,y2),(0,255,0), 2)   

# Different weights are added to the image to give a feeling of blending
def blend_images(image, final_image, alpha=0.7, beta=1., gamma=0.):
    return cv2.addWeighted(final_image, alpha, image, beta,gamma)

image = cv2.imread("road.jpg") # load image in grayscale
gray_image = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
blurredImage = cv2.GaussianBlur(gray_image, (5, 5), 0)
edgeImage = cv2.Canny(blurredImage, 50, 120)

# Detect points that form a line
dis_reso = 1 # Distance resolution in pixels of the Hough grid
theta = np.pi /180 # Angular resolution in radians of the Hough grid
threshold = 170# minimum no of votes

houghLines = cv2.HoughLines(edgeImage, dis_reso, theta, threshold)

houghLinesImage = np.zeros_like(image) # create and empty image

drawhoughLinesOnImage(houghLinesImage, houghLines) # draw the lines on the empty image
orginalImageWithHoughLines = blend_images(houghLinesImage,image) # add two images together, using image blending



cv2.imshow("Original image", image)
cv2.waitKey(0)

cv2.imshow("Edge Image'", edgeImage)
cv2.waitKey(0)

cv2.imshow("Original Image with Hough lines", orginalImageWithHoughLines)
cv2.waitKey(0)

cv2.destroyAllWindows();

File: DIP/DIP_009/C++/DIP_009_line_detector.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

int main()
{
    // Declare the output variables
    Mat edgeImage, orginalImageWithHoughLines;
    
    // Loads an image
    Mat image = imread( "road.jpg", IMREAD_GRAYSCALE);

    // Check if image is loaded fine
    if(image.empty()){
        printf(" Error opening imagen");
       return -1;
    }

    // Apply the Guassian Blur filter to smooth the image
    cv::Mat image_gaussian_processed;
    cv::GaussianBlur(image, image_gaussian_processed, Size(5,5), 1);

    // Edge detection
    Canny(image_gaussian_processed, edgeImage, 50, 120, 3);

    // Copy loaded image to the initial image so that will display the results in BGR
    cvtColor(image, orginalImageWithHoughLines, COLOR_GRAY2BGR);

    // Declaring some constants for the parameters
    int dis_reso = 1;
    double theta = CV_PI/180;
    int threshold = 170;

    // Standard Hough Line Transform
    vector<Vec2f> lines; // will hold the results of the detection
    HoughLines(edgeImage, lines, dis_reso, theta, threshold, 0, 0); // runs the actual detection

    // Draw the lines
    for( size_t i = 0; i < lines.size(); i++ )
    {
        float rho = lines[i][0], theta = lines[i][1];
        Point pt1, pt2;
        double a = cos(theta), b = sin(theta);
        double x0 = a*rho, y0 = b*rho;
        pt1.x = cvRound(x0 + 1000*(-b));
        pt1.y = cvRound(y0 + 1000*(a));
        pt2.x = cvRound(x0 - 1000*(-b));
        pt2.y = cvRound(y0 - 1000*(a));
        line( orginalImageWithHoughLines, pt1, pt2, Scalar(0,255,00), 3, LINE_AA);
    }

    //cvtColor(image, image, COLOR_GRAY2BGR);
    //cvtColor(orginalImageWithHoughLines, orginalImageWithHoughLines, COLOR_GRAY2BGR);

    // Show results
    cv::imshow("Original image", image);
    cv::imshow("Detected Lines (in Green) - Standard Hough Line Transform", orginalImageWithHoughLines);
    // Wait and Exit
    waitKey(0);

    return 0;
}